File: app/nehex-admin/src/utils/commentTargets.ts
type CommentTargetLike = {
  target_type: string
  target_id: number
}

type CommentLinkLike = CommentTargetLike & {
  id: number
}

type StandalonePageResolver = (pageId: number) => Promise<string | null | undefined>

export function normalizeCommentTargetType(targetType: string): string {
  return String(targetType || '').trim().toLowerCase()
}

export function mapCommentTargetLabel(targetType: string): string {
  const normalizedType = normalizeCommentTargetType(targetType)
  const labelMap: Record<string, string> = {
    article: '文章',
    album: '相册',
    daily: '日常',
    project: '项目',
    singlepage: '独立页',
    friend_page: '友链页',
  }
  return labelMap[normalizedType] || normalizedType || '未知'
}

export function canJumpToCommentTarget(comment: CommentTargetLike): boolean {
  const targetType = normalizeCommentTargetType(comment.target_type)
  if (targetType === 'friend_page') {
    return true
  }

  const targetId = Number(comment.target_id)
  if (!Number.isFinite(targetId) || targetId <= 0) {
    return false
  }

  return targetType === 'article'
    || targetType === 'album'
    || targetType === 'daily'
    || targetType === 'project'
    || targetType === 'singlepage'
}

export function buildCommentTargetPath(
  targetType: string,
  targetId: number,
  pageKey?: string | null,
): string {
  const normalizedType = normalizeCommentTargetType(targetType)
  const normalizedId = Math.max(1, Math.floor(Number(targetId) || 0))

  if (normalizedType === 'article') {
    return `/article/${normalizedId}`
  }
  if (normalizedType === 'album') {
    return `/album/${normalizedId}`
  }
  if (normalizedType === 'daily') {
    return `/daily/${normalizedId}`
  }
  if (normalizedType === 'project') {
    return `/project/${normalizedId}`
  }
  if (normalizedType === 'friend_page') {
    return '/friends'
  }
  if (normalizedType === 'singlepage') {
    const normalizedPageKey = String(pageKey ?? '').trim().replace(/^\/+|\/+$/g, '')
    return normalizedPageKey ? `/${normalizedPageKey}` : `/page/${normalizedId}`
  }
  return ''
}

export async function resolveCommentTargetPath(
  comment: CommentTargetLike,
  resolveStandalonePagePath?: StandalonePageResolver,
): Promise<string> {
  const targetType = normalizeCommentTargetType(comment.target_type)
  const targetId = Math.max(1, Math.floor(Number(comment.target_id) || 0))

  if (targetType !== 'singlepage') {
    return buildCommentTargetPath(targetType, targetId)
  }

  if (resolveStandalonePagePath) {
    const pagePath = await resolveStandalonePagePath(targetId)
    if (pagePath) {
      return buildCommentTargetPath(targetType, targetId, pagePath)
    }
  }

  return buildCommentTargetPath(targetType, targetId)
}

export function joinSiteUrl(baseUrl: string, path: string): string {
  const normalizedPath = path.trim() ? `/${path.trim().replace(/^\/+/, '')}` : '/'
  const normalizedBase = baseUrl.trim().replace(/\/+$/, '')
  return normalizedBase ? `${normalizedBase}${normalizedPath}` : normalizedPath
}

export function withCommentAnchor(url: string, commentId: number): string {
  const normalizedUrl = String(url || '').trim()
  if (!normalizedUrl) {
    return ''
  }
  return `${normalizedUrl}#comment-${Math.max(1, Math.floor(Number(commentId) || 0))}`
}

export async function buildCommentTargetUrl(
  comment: CommentLinkLike,
  siteUrl: string,
  resolveStandalonePagePath?: StandalonePageResolver,
): Promise<string> {
  const path = await resolveCommentTargetPath(comment, resolveStandalonePagePath)
  return withCommentAnchor(joinSiteUrl(siteUrl, path), comment.id)
}

File: backend-rust/src/routes/comment_targets.rs
pub fn normalize_comment_target_type(target_type: &str) -> String {
    target_type.trim().to_lowercase()
}

pub fn build_comment_sync_token(target_type: &str, target_id: i64) -> String {
    format!(
        "{}:{}",
        normalize_comment_target_type(target_type),
        target_id.max(1)
    )
}

pub fn build_comment_target_path(
    target_type: &str,
    target_id: i64,
    page_key: Option<&str>,
) -> String {
    let normalized_type = normalize_comment_target_type(target_type);
    let normalized_id = target_id.max(1);

    match normalized_type.as_str() {
        "article" => format!("/article/{normalized_id}"),
        "album" => format!("/album/{normalized_id}"),
        "daily" => format!("/daily/{normalized_id}"),
        "project" => format!("/project/{normalized_id}"),
        "friend_page" => "/friends".to_string(),
        "singlepage" => {
            let normalized_page_key = page_key
                .map(|value| value.trim().trim_matches('/'))
                .filter(|value| !value.is_empty());
            if let Some(value) = normalized_page_key {
                format!("/{value}")
            } else {
                format!("/page/{normalized_id}")
            }
        }
        _ => "/".to_string(),
    }
}

pub fn join_site_url(site_url: &str, path: &str) -> String {
    let normalized_path = if path.trim().is_empty() {
        "/".to_string()
    } else {
        format!("/{}", path.trim().trim_start_matches('/'))
    };

    if site_url.trim().is_empty() {
        normalized_path
    } else {
        format!(
            "{}{}",
            site_url.trim().trim_end_matches('/'),
            normalized_path
        )
    }
}

pub fn with_comment_anchor(target_url: &str, comment_id: i64) -> String {
    if target_url.trim().is_empty() {
        String::new()
    } else {
        format!("{}#comment-{}", target_url, comment_id.max(1))
    }
}

#[cfg(test)]
mod tests {
    use super::{
        build_comment_sync_token, build_comment_target_path, join_site_url, with_comment_anchor,
    };

    #[test]
    fn comment_sync_token_is_normalized() {
        assert_eq!(build_comment_sync_token(" Daily ", 0), "daily:1");
        assert_eq!(build_comment_sync_token("PROJECT", 42), "project:42");
    }

    #[test]
    fn comment_target_path_covers_public_routes() {
        assert_eq!(build_comment_target_path("article", 7, None), "/article/7");
        assert_eq!(build_comment_target_path("album", 8, None), "/album/8");
        assert_eq!(build_comment_target_path("daily", 9, None), "/daily/9");
        assert_eq!(build_comment_target_path("project", 10, None), "/project/10");
        assert_eq!(
            build_comment_target_path("friend_page", 0, None),
            "/friends"
        );
    }

    #[test]
    fn singlepage_target_path_prefers_page_key() {
        assert_eq!(
            build_comment_target_path("singlepage", 11, Some(" about/me ")),
            "/about/me"
        );
        assert_eq!(build_comment_target_path("singlepage", 11, Some("///")), "/page/11");
        assert_eq!(build_comment_target_path("singlepage", 11, None), "/page/11");
    }

    #[test]
    fn comment_target_url_helpers_build_full_direct_link() {
        let url = join_site_url("https://example.com/", "/daily/12");
        assert_eq!(url, "https://example.com/daily/12");
        assert_eq!(
            with_comment_anchor(&url, 34),
            "https://example.com/daily/12#comment-34"
        );
    }
}
